src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -3888,61 +3905,70 @@ private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisp
+                                          : _compilation.GetSpecialType(SpecialType.System_IDisposable);
 
-            AddStatement(tryDispose(resource) ??
-                         MakeInvalidOperation(type: null, resource));
+                var isPatternDispose = !disposeMethod.ContainingType.Equals(iDisposable, SymbolEqualityComparer.IgnoreAll);
 
-            AppendNewBlock(endOfFinally);
-
-            Debug.Assert(_currentRegion == finallyRegion);
-            LeaveRegion();
-            return;
+                // var resource = resource as IDisposable;
+                if (requiresRuntimeConversion)
+                {
+                    Debug.Assert(!isNonNullableValueType(resource.Type));
+                    resource = ConvertToIDisposable(resource, iDisposable, isTryCast: true);
+                    int captureId = GetNextCaptureId(finallyRegion);
+                    AddStatement(new FlowCaptureOperation(captureId, resource.Syntax, resource));
+                    resource = GetCaptureReference(captureId, resource);
+                }
 
-            IOperation? tryDispose(IOperation value)
-            {
-                Debug.Assert(value.Type!.Equals(iDisposable));
+                // if (resource is null) return;
+                if (requiresRuntimeConversion || !isNonNullableValueType(resource.Type))
+                {
+                    IOperation condition = MakeIsNullOperation(OperationCloner.CloneOperation(resource));
+                    ConditionalBranch(condition, jumpIfTrue: true, endOfFinally);
+                    _currentBasicBlock = null;
+                }
 
-                var method = isAsynchronous
-                    ? (IMethodSymbol?)_compilation.CommonGetWellKnownTypeMember(WellKnownMember.System_IAsyncDisposable__DisposeAsync)?.GetISymbol()
-                    : (IMethodSymbol?)_compilation.CommonGetSpecialTypeMember(SpecialMember.System_IDisposable__Dispose)?.GetISymbol();
-                if (method != null)
+                // ((IDisposable)resource)
+                if (!isPatternDispose && !iDisposable.Equals(resource.Type))
                 {
-                    var invocation = new InvocationOperation(method, value, isVirtual: true,
-                                                             ImmutableArray<IArgumentOperation>.Empty, semanticModel: null, value.Syntax,
-                                                             method.ReturnType, isImplicit: true);
+                    resource = ConvertToIDisposable(resource, iDisposable);
+                }
 
-                    if (isAsynchronous)
-                    {
-                        return new AwaitOperation(invocation, semanticModel: null, value.Syntax, _compilation.GetSpecialType(SpecialType.System_Void), isImplicit: true);
-                    }
+                // resource.Dispose();
+                resource = new InvocationOperation(disposeMethod, resource, isVirtual: true,
```

Comment:
>new InvocationOperation(disposeMethod, resource, isVirtual: true, [](start = 27, length = 65)

Can pattern-based expose use an extension method? In that scenario, the resource should be the first parameter instead. Do we need to handle omitted optional and param array parameters?

Reply:
No, pattern disposal does not support extensions methods. We should report the params and optional arguments, however

Reply:
> We should report the params and optional arguments, however

See how we deal with arguments for GetEnumerator and MoveNext in VisitForEachLoop

---
In reply to: [533774968](https://github.com/dotnet/roslyn/pull/49481#discussion_r533774968) [](ancestors = 533774968)

Reply:
@AlekseyTs I have a fix for this, but its a little involved, as we now need to pass the binder down through the `BoundUsingStatement` like we do for Foreach. Given the mild complexity, do you want me to add it to this PR, or wait and open a separate one?

Reply:
> Given the mild complexity, do you want me to add it to this PR, or wait and open a separate one?

I think a separate PR is fine.

---
In reply to: [534633638](https://github.com/dotnet/roslyn/pull/49481#discussion_r534633638) [](ancestors = 534633638)